src/appleseed/renderer/modeling/aov/normalaov.cpp

```diff
@@ -0,0 +1,190 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Esteban Tovagliari, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "normalaov.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+#include "renderer/kernel/aov/aovaccumulator.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingresult.h"
+#include "renderer/modeling/aov/aov.h"
+#include "renderer/utility/paramarray.h"
+
+// appleseed.foundation headers.
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// Standard headers.
+#include <string>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Normal AOV accumulator.
+    //
+
+    class NormalAOVAccumulator
+      : public AOVAccumulator
+    {
+      public:
+        explicit NormalAOVAccumulator(const size_t index)
+          : AOVAccumulator(index)
+        {
+        }
+
+        virtual void reset() APPLESEED_OVERRIDE
+        {
+            m_normal = Vector3f(0.0f, 0.0f, 0.0f);
+        }
+
+        virtual void write(
+            const ShadingPoint&     shading_point,
+            const Camera&           camera) APPLESEED_OVERRIDE
+        {
+            if (shading_point.hit())
+                m_normal = Vector3f(shading_point.get_shading_normal());
+        }
+
+        virtual void flush(ShadingResult& result) APPLESEED_OVERRIDE
+        {
+            result.m_aovs[m_index].m_color =
```

Comment:
It could happen because of numerical imprecisions. Since we know here that it cannot happen, a saturate call would make sure there's no problem downstream.